L82-core-event-engine.md

```diff
@@ -0,0 +1,478 @@
+* Client logic, support for making outgoing network connections.
+
+### Common Data Types
+
+```cpp
+namespace grpc_event_engine {
+namespace experimental {
+class EventEngine {
+ public:
+  /// A basic callable function. The first argument to all callbacks is an
+  /// absl::Status indicating the status of the operation associated with this
+  /// callback. Each EventEngine method that takes a callback parameter, defines
+  /// the expected sets and meanings of statuses for that use case.
+  using Callback = std::function<void(absl::Status)>;
+  /// A callback handle, used to cancel a callback.
+  struct TaskHandle {
+    intptr_t key;
+  };
+  /// A thin wrapper around a platform-specific sockaddr type. A sockaddr struct
+  /// exists on all platforms that gRPC supports.
+  ///
+  /// Platforms are expected to provide definitions for:
+  /// * sockaddr
+  /// * sockaddr_in
+  /// * sockaddr_in6
+  class ResolvedAddress {
+   public:
+    static constexpr socklen_t MAX_SIZE_BYTES = 128;
+
+    ResolvedAddress(const sockaddr* address, socklen_t size);
+    ResolvedAddress() = default;
+    ResolvedAddress(const ResolvedAddress&) = default;
+    const struct sockaddr* address() const;
+    socklen_t size() const;
+
+   private:
+    char address_[MAX_SIZE_BYTES];
+    socklen_t size_ = 0;
+  };
+  virtual ~EventEngine() = default;
+```
+
+### Endpoints
+
+```cpp
+  /// An Endpoint represents one end of a connection between a gRPC client and
+  /// server. Endpoints are created when connections are established, and
+  /// Endpoint operations are gRPC's primary means of communication.
+  ///
+  /// Endpoints must use the provided SliceAllocator for all data buffer memory
+  /// allocations. gRPC allows applications to set memory constraints per
+  /// Channel or Server, and the implementation depends on all dynamic memory
+  /// allocation being handled by the quota system.
+  class Endpoint {
+   public:
+    /// The Endpoint destructor is responsible for shutting down all connections
+    /// and invoking all pending read or write callbacks with an error status.
+    virtual ~Endpoint() = default;
+    /// Read data from the Endpoint.
+    ///
+    /// When data is available on the connection, that data is moved into the
+    /// \a buffer, and the \a on_read callback is called. The caller must ensure
+    /// that the callback has access to the buffer when executed later.
+    /// Ownership of the buffer is not transferred. Valid slices *may* be placed
+    /// into the buffer even if the callback is invoked with a non-OK Status.
+    ///
+    /// For failed read operations, implementations should pass the appropriate
+    /// statuses to \a on_read. For example, callbacks might expect to receive
+    /// DEADLINE_EXCEEDED when the deadline is exceeded, and CANCELLED on
+    /// endpoint shutdown.
+    virtual void Read(Callback on_read, SliceBuffer* buffer,
+                      absl::Time deadline) = 0;
```

Comment:
Please explain the rationale for adding a timeout to read?

Reply:
I believe @markdroth and Sanjay were the key proponents of it. Care to weigh in, Mark?

Reply:
Just to check my assumptions while I wait for that: my belief is that it's uncommon for operating systems to provide a primitive that will perform an asynchronous read with a deadline, and so we're forcing all (or at least most) implementations to emulate it. Further, most reads in gRPC don't need a deadline, so it looks we're forcing adding functionality we don't need (and the associated performance cost), as well as implementation complexity to every event engine that's unwarranted.

Reply:
I think we agreed on Friday to remove the deadline... Thinking about this some more, should this return a ReadTaskHandle so that we can offer a cancellation mechanism? (Maybe we can do that later)

Should it take a ReadOptions akin to RunOptions for future API expansion?

Reply:
Correct, I've removed the deadlines from the API. I didn't find any discussion on the deadlines at all, going back to early reviews, so I think they were at most hypothetically useful from the beginning.

Regarding `ReadOptions`, that sounds potentially useful. Can you think of any way it would be used that couldn't otherwise be solved by writing a custom EventEngine?

For `RunOptions`, we had discussed maybe needing a `priority` field to be added for efficiency of some implementations. But unless gRPC-internals would use the feature (currently, there is no run-priority concept internally), then `priority` would have been added speculatively at this point.

Reply:
ReadOptions I was wondering about the urgent flag we currently have... I have no deep insight into it's purpose, but at some point we felt like we needed it, and maybe we'll want such flexibility again (actually akin to priority for RunOptions, but I think it would be a mistake to tie the concepts!)

Reply:
With regard to read cancellation, I think the current plan is that the only time a read would need to be cancelled would be if the endpoint is shutting down.  I don't think we have any cases where we need to cancel a single read without shutting down the endpoint.  But if you have such a case in mind, we can discuss.

Reply:
I don't... I think I had an API symmetry kick last week.

That said: I think if we resolve RunOptions as wanting to keep it, we should probably add ReadOptions, and similarly if we decide RunOptions is not the right thing we should not add ReadOptions.

Reply:
Closing the loop: we removed the `*Options` structs as YAGNI. The API is experimental and can be modified for a bit. We may introduce these structs later if we have cases for them.